File: src/main/java/de/danielmaile/aether/mobs/RideableLlama.java
package de.danielmaile.aether.mobs;

import com.comphenix.protocol.PacketType;
import com.comphenix.protocol.ProtocolManager;
import com.comphenix.protocol.events.ListenerPriority;
import com.comphenix.protocol.events.PacketAdapter;
import com.comphenix.protocol.events.PacketContainer;
import com.comphenix.protocol.events.PacketEvent;
import de.danielmaile.aether.Aether;
import de.danielmaile.aether.worldgen.AetherWorld;
import org.bukkit.entity.Entity;
import org.bukkit.entity.Llama;
import org.bukkit.entity.Player;
import org.bukkit.util.Vector;

public class RideableLlama
{
    public void init(ProtocolManager protocolManager)
    {
        protocolManager.addPacketListener(new PacketAdapter(Aether.getInstance(), ListenerPriority.NORMAL, PacketType.Play.Client.STEER_VEHICLE)
        {
            @Override
            public void onPacketReceiving(PacketEvent event)
            {
                PacketContainer packetContainer = event.getPacket();
                Player player = event.getPlayer();
                Entity vehicle = player.getVehicle();

                if(vehicle == null) return;
                if(!(vehicle instanceof Llama llama)) return;
                if(!llama.getLocation().getWorld().equals(AetherWorld.getWorld())) return;
                if(llama.getInventory().getDecor() == null) return;

                //Jumping
                boolean jump = packetContainer.getBooleans().read(0);
                if(jump && llama.isOnGround())
                {
                    llama.setVelocity(llama.getVelocity().add(new Vector(0.0, Aether.getInstance().getConfigManager().getLlamaJumpHeight(), 0.0)));
                }

                //Calculate velocity
                float leftRight = packetContainer.getFloat().read(0);
                float forwardBackward = packetContainer.getFloat().read(1);

                Vector horizontal = new Vector(-forwardBackward, 0, leftRight);
                if (horizontal.length() > 0)
                {
                    //Turn to face of vector in direction which the player is facing
                    horizontal.rotateAroundY(Math.toRadians(-player.getLocation().getYaw() + 90f));

                    //Scale vector
                    horizontal.normalize().multiply(Aether.getInstance().getConfigManager().getLlamaSpeed());
                }

                Vector vertical = llama.getVelocity();
                vertical.setX(0);
                vertical.setZ(0);
                llama.setVelocity(horizontal.add(vertical));

                //Rotation
                llama.setRotation(player.getEyeLocation().getYaw(), player.getEyeLocation().getPitch());
            }
        });
    }
}

File: src/main/java/de/danielmaile/aether/mobs/RideableLlama.kt
package de.danielmaile.aether.mobs

import com.comphenix.protocol.PacketType
import com.comphenix.protocol.ProtocolManager
import com.comphenix.protocol.events.ListenerPriority
import com.comphenix.protocol.events.PacketAdapter
import com.comphenix.protocol.events.PacketEvent
import de.danielmaile.aether.inst
import org.bukkit.entity.Llama
import org.bukkit.util.Vector

class RideableLlama(protocolManager: ProtocolManager) {
    init {
        protocolManager.addPacketListener(object :
            PacketAdapter(inst(), ListenerPriority.NORMAL, PacketType.Play.Client.STEER_VEHICLE) {
            override fun onPacketReceiving(event: PacketEvent) {
                val packetContainer = event.packet
                val player = event.player
                val llama = player.vehicle ?: return

                if (llama !is Llama) return
                if (llama.location.world != inst().aetherWorld.getWorld()) return
                if (llama.inventory.decor == null) return

                //Jumping
                val jump = packetContainer.booleans.read(0)
                if (jump && llama.isOnGround()) {
                    llama.velocity = llama.velocity.clone().add(Vector(0.0, inst().configManager.llamaJumpHeight, 0.0))
                }

                //Calculate velocity
                val leftRight = packetContainer.float.read(0)
                val forwardBackward = packetContainer.float.read(1)

                val horizontal = Vector(-forwardBackward.toDouble(), 0.0, leftRight.toDouble())
                if (horizontal.length() > 0) {
                    //Turn to face of vector in direction which the player is facing
                    horizontal.rotateAroundY(Math.toRadians((-player.location.yaw + 90f).toDouble()))

                    //Scale vector
                    horizontal.normalize().multiply(inst().configManager.llamaSpeed)
                }

                val vertical = llama.getVelocity()
                vertical.setX(0)
                vertical.setZ(0)
                llama.setVelocity(horizontal.add(vertical))

                //Rotation
                llama.setRotation(player.eyeLocation.yaw, player.eyeLocation.pitch)
            }
        })
    }
}